File: sdks/node/tests/git.integration.test.ts
/**
 * Integration tests for SimpleBox.git.
 *
 * Same contract as the Python and Go git integration tests: configure_user /
 * set_config / get_config reach guest git config through the native binding.
 * Writes go through box.git; observations go through guest `git config` /
 * `git log` (or the reverse for getConfig).
 */

import { describe, test, expect, beforeAll, afterAll } from "vitest";
import { SimpleBox } from "../lib/simplebox.js";

describe("git config namespace", { timeout: 120_000 }, () => {
  let box: SimpleBox;
  let gitInstalled = false;

  beforeAll(async () => {
    box = new SimpleBox({ image: "alpine:latest" });
    const warmup = await box.exec("true");
    expect(warmup.exitCode).toBe(0);

    const installed = await box.exec("sh", "-c", "apk add --no-cache git");
    gitInstalled = installed.exitCode === 0;
    if (!gitInstalled) {
      console.warn(`apk add git failed: ${installed.stderr}`);
    }
  });

  afterAll(async () => {
    await box.stop();
  });

  test("local scope without path names the missing argument", async () => {
    await expect(
      box.git.setConfig("user.email", "local@boxlite.ai", "local"),
    ).rejects.toThrow(/path/);
    await expect(box.git.getConfig("user.email", "local")).rejects.toThrow(
      /path/,
    );
    await expect(
      box.git.configureUser("BoxLite Bot", "bot@boxlite.ai", "local"),
    ).rejects.toThrow(/path/);
  });

  test("unknown scope is rejected with the allowed values", async () => {
    await expect(box.git.setConfig("user.email", "x", "file")).rejects.toThrow(
      /global/,
    );
    await expect(box.git.getConfig("user.email", "file")).rejects.toThrow(
      /global/,
    );
    await expect(
      box.git.configureUser("BoxLite Bot", "x@boxlite.ai", "file"),
    ).rejects.toThrow(/global/);
  });

  test("configureUser writes global identity into the guest", async ({
    skip,
  }) => {
    if (!gitInstalled) {
      skip("apk add git failed");
    }

    await box.git.configureUser("BoxLite Bot", "bot@boxlite.ai");
    const email = await box.exec(
      "sh",
      "-c",
      "git config --global --get user.email",
    );
    expect(email.exitCode).toBe(0);
    expect(email.stdout.trim()).toBe("bot@boxlite.ai");
    const name = await box.exec(
      "sh",
      "-c",
      "git config --global --get user.name",
    );
    expect(name.exitCode).toBe(0);
    expect(name.stdout.trim()).toBe("BoxLite Bot");
  });

  test("commit records the configured author", async ({ skip }) => {
    if (!gitInstalled) {
      skip("apk add git failed");
    }

    await box.git.configureUser("BoxLite Bot", "bot@boxlite.ai");
    const init = await box.exec(
      "sh",
      "-c",
      [
        "set -e",
        "git init /tmp/git-commit",
        "echo hi > /tmp/git-commit/README",
        "git -C /tmp/git-commit add README",
        "git -C /tmp/git-commit -c commit.gpgsign=false commit -m init",
        "git -C /tmp/git-commit log -1 --format='%an <%ae>'",
      ].join("\n"),
    );
    expect(init.exitCode).toBe(0);
    expect(init.stdout).toContain("BoxLite Bot <bot@boxlite.ai>");
  });

  test("getConfig reads guest global config", async ({ skip }) => {
    if (!gitInstalled) {
      skip("apk add git failed");
    }

    const written = await box.exec(
      "sh",
      "-c",
      "git config --global user.email other@boxlite.ai",
    );
    expect(written.exitCode).toBe(0);
    expect(await box.git.getConfig("user.email")).toBe("other@boxlite.ai");
    await box.git.configureUser("BoxLite Bot", "bot@boxlite.ai");
  });

  test("local setConfig does not change global", async ({ skip }) => {
    if (!gitInstalled) {
      skip("apk add git failed");
    }

    await box.git.configureUser("BoxLite Bot", "bot@boxlite.ai");
    const init = await box.exec("git", "init", "/tmp/git-local");
    expect(init.exitCode).toBe(0);

    await box.git.setConfig(
      "user.email",
      "local@boxlite.ai",
      "local",
      "/tmp/git-local",
    );
    const local = await box.exec(
      "sh",
      "-c",
      "git -C /tmp/git-local config --local --get user.email",
    );
    expect(local.exitCode).toBe(0);
    expect(local.stdout.trim()).toBe("local@boxlite.ai");
    expect(
      await box.git.getConfig("user.email", "local", "/tmp/git-local"),
    ).toBe("local@boxlite.ai");
    const globalEmail = await box.exec(
      "sh",
      "-c",
      "git config --global --get user.email",
    );
    expect(globalEmail.exitCode).toBe(0);
    expect(globalEmail.stdout.trim()).toBe("bot@boxlite.ai");
    expect(await box.git.getConfig("user.email")).toBe("bot@boxlite.ai");
  });
});

File: sdks/go/git_integration_test.go
//go:build boxlite_dev

package boxlite

import (
	"bytes"
	"context"
	"errors"
	"strings"
	"testing"
)

// TestIntegrationGitConfig proves that ConfigureUser / SetConfig / GetConfig
// actually reach guest git config — i.e. the Go SDK's git handle plumbing
// makes it through the C FFI and out the other side. A single box is reused
// for all checks because creating a VM dominates the test cost.
//
// Each subtest asserts a project-symbol path:
//   - ConfigureUser: Git.ConfigureUser -> boxlite_git_configure_user ->
//     GitHandle::configure_user -> git config --global
//   - GetConfig:     Git.GetConfig -> boxlite_git_get_config ->
//     GitHandle::get_config -> git config --get
//   - SetConfig local: Git.SetConfig(scope=local, path) -> working_dir ->
//     git config --local
//   - GetConfig local: Git.GetConfig(scope=local, path) -> git config --get
//   - Missing path:  Git.SetConfig(scope=local) -> InvalidArgument (no guest exec)
func TestIntegrationGitConfig(t *testing.T) {
	rt := newTestRuntime(t)
	box := createStartedBoxOrSkip(
		t,
		rt,
		"alpine:latest",
		WithNetwork(NetworkSpec{
			Outbound: OutboundNetworkSpec{Mode: NetworkModeEnabled},
		}),
		WithAutoRemove(false),
	)
	installGitOrSkip(t, box)

	ctx := context.Background()
	git, err := box.Git()
	if err != nil {
		t.Fatalf("Git(): %v", err)
	}
	t.Cleanup(func() { _ = git.Close() })

	// Short-lived `git config --get` can exit so fast that Wait returns
	// before the stdout pump delivers the last bytes. Same pad as
	// TestIntegrationExecEnvWorkingDirTimeout.
	const drainPad = " && sleep 0.1"

	t.Run("ConfigureUser reaches guest global config", func(t *testing.T) {
		if err := git.ConfigureUser(ctx, "BoxLite Bot", "bot@boxlite.ai", nil); err != nil {
			t.Fatalf("ConfigureUser: %v", err)
		}
		email := guestStdout(t, box, "git config --global --get user.email"+drainPad)
		if email != "bot@boxlite.ai" {
			t.Fatalf("user.email did not reach guest: want bot@boxlite.ai, got %q", email)
		}
		name := guestStdout(t, box, "git config --global --get user.name"+drainPad)
		if name != "BoxLite Bot" {
			t.Fatalf("user.name did not reach guest: want BoxLite Bot, got %q", name)
		}

		log := guestStdout(t, box, "set -e\n"+
			"git init /tmp/repo\n"+
			"echo hi > /tmp/repo/README\n"+
			"git -C /tmp/repo add README\n"+
			"git -C /tmp/repo -c commit.gpgsign=false commit -m init\n"+
			"git -C /tmp/repo log -1 --format='%an <%ae>'"+drainPad)
		if !strings.Contains(log, "BoxLite Bot <bot@boxlite.ai>") {
			t.Fatalf("commit did not record configured author: got %q", log)
		}
	})

	t.Run("GetConfig reads guest global config", func(t *testing.T) {
		set := box.Command("sh", "-c", "git config --global user.email other@boxlite.ai")
		if err := set.Run(ctx); err != nil {
			t.Fatalf("git config --global user.email: %v", err)
		}
		email, err := git.GetConfig(ctx, "user.email", nil)
		if err != nil {
			t.Fatalf("GetConfig: %v", err)
		}
		if email != "other@boxlite.ai" {
			t.Fatalf("GetConfig did not read guest: want other@boxlite.ai, got %q", email)
		}
		if err := git.ConfigureUser(ctx, "BoxLite Bot", "bot@boxlite.ai", nil); err != nil {
			t.Fatalf("restore ConfigureUser: %v", err)
		}
	})

	t.Run("local scope without path is InvalidArgument", func(t *testing.T) {
		err := git.SetConfig(ctx, "user.email", "local@boxlite.ai", &GitConfigOptions{
			Scope: "local",
		})
		if err == nil {
			t.Fatal("SetConfig(local, no path) succeeded, want InvalidArgument")
		}
		var boxliteErr *Error
		if !errors.As(err, &boxliteErr) || boxliteErr.Code != ErrInvalidArgument {
			t.Fatalf("error = %v, want InvalidArgument", err)
		}
		if !strings.Contains(boxliteErr.Message, "path") {
			t.Fatalf("error %q should name path", boxliteErr.Message)
		}
	})

	t.Run("SetConfig local reaches that repo only", func(t *testing.T) {
		init := box.Command("git", "init", "/tmp/repo")
		if err := init.Run(ctx); err != nil {
			t.Fatalf("git init: %v", err)
		}
		if err := git.SetConfig(ctx, "user.email", "local@boxlite.ai", &GitConfigOptions{
			Scope: "local",
			Path:  "/tmp/repo",
		}); err != nil {
			t.Fatalf("SetConfig local: %v", err)
		}
		localEmail := guestStdout(t, box, "git -C /tmp/repo config --local --get user.email"+drainPad)
		if localEmail != "local@boxlite.ai" {
			t.Fatalf("local user.email did not reach guest: want local@boxlite.ai, got %q", localEmail)
		}
		gotLocal, err := git.GetConfig(ctx, "user.email", &GitConfigOptions{
			Scope: "local",
			Path:  "/tmp/repo",
		})
		if err != nil {
			t.Fatalf("GetConfig local: %v", err)
		}
		if gotLocal != "local@boxlite.ai" {
			t.Fatalf("GetConfig local = %q, want local@boxlite.ai", gotLocal)
		}
		globalEmail := guestStdout(t, box, "git config --global --get user.email"+drainPad)
		if globalEmail != "bot@boxlite.ai" {
			t.Fatalf("global user.email changed: want bot@boxlite.ai, got %q", globalEmail)
		}
		gotGlobal, err := git.GetConfig(ctx, "user.email", nil)
		if err != nil {
			t.Fatalf("GetConfig global: %v", err)
		}
		if gotGlobal != "bot@boxlite.ai" {
			t.Fatalf("GetConfig global = %q, want bot@boxlite.ai", gotGlobal)
		}
	})
}

func guestStdout(t *testing.T, box *Box, script string) string {
	t.Helper()
	cmd := box.Command("sh", "-c", script)
	var out, stderr bytes.Buffer
	cmd.Stdout = &out
	cmd.Stderr = &stderr
	if err := cmd.Run(context.Background()); err != nil {
		t.Fatalf("guest %q: %v: %s", script, err, stderr.String())
	}
	return strings.TrimSpace(out.String())
}

func installGitOrSkip(t *testing.T, box *Box) {
	t.Helper()
	cmd := box.Command("sh", "-c", "apk add --no-cache git")
	var stderr bytes.Buffer
	cmd.Stderr = &stderr
	if err := cmd.Run(context.Background()); err != nil {
		t.Skipf("apk add git: %v: %s", err, stderr.String())
	}
}
